src/order/initial_seg.lean

```diff
@@ -107,19 +107,18 @@ end⟩
 
 @[simp] theorem trans_apply (f : r ≼i s) (g : s ≼i t) (a : α) : (f.trans g) a = g (f a) := rfl
 
-theorem unique_of_trichotomous_of_irrefl [is_trichotomous β s] [is_irrefl β s] :
-  well_founded r → subsingleton (r ≼i s) | ⟨h⟩ :=
+instance subsingleton_of_trichotomous_of_irrefl [is_trichotomous β s] [is_irrefl β s]
+  [is_well_founded α r] : subsingleton (r ≼i s) :=
 ⟨λ f g, begin
   ext a,
-  have := h a, induction this with a H IH,
+  apply is_well_founded.induction r a (λ b IH, _),
   refine extensional_of_trichotomous_of_irrefl s (λ x, _),
-  simp only [f.init_iff, g.init_iff],
+  rw [f.init_iff, g.init_iff],
   exact exists_congr (λ x, and_congr_left $ λ hx, IH _ hx ▸ iff.rfl)
 end⟩
 
 instance [is_well_order β s] : subsingleton (r ≼i s) :=
-⟨λ a, @subsingleton.elim _ (unique_of_trichotomous_of_irrefl
-  (@rel_embedding.well_founded _ _ r s a is_well_founded.wf)) a⟩
+⟨λ a, by { letI := a.is_well_founded, apply subsingleton.elim }⟩
```

Comment:
Doesn't this work?
```suggestion
⟨λ a, subsingleton.elim _ _⟩
```

Reply:
`a.is_well_founded` can't be inferred since it makes use of `a` as data.